Add test for specialization constants defined various ways

vasilytric opened the pull request:
Provides tests for specialization constants defined various ways

Based on #140  (got common code (tests/specialization_constants/specialization_constants_common.h) from this PR)

Please review from commit: 9a6354a6fd0439433f928452250c5bfeb841b0af

keryell commented:
I compared directly `kovshikovser/upstream_spec_const/part1..vasilytric/specialization_constants_defined_various_ways` but it is difficult to make progress on this and the following PR while https://github.com/KhronosGroup/SYCL-CTS/pull/140 has not landed because there are some copy paste from https://github.com/KhronosGroup/SYCL-CTS/pull/140 that should be updated accordingly.

bader commented:
@keryell here is the link on the changes to review within this PR: https://github.com/KhronosGroup/SYCL-CTS/commit/9a6354a6fd0439433f928452250c5bfeb841b0af.
This PR has 4 commits:
- first two commits are from https://github.com/KhronosGroup/SYCL-CTS/pull/140 and can be ignored in the scope of this pull request.
- merge commit to bring previous commits to the branch - ignore
- 9a6354a6fd0439433f928452250c5bfeb841b0af is the commit for review.

bader commented:
@keryell, @vasilytric, @kovshikovser, do you think it make sense to create a single pull request for specialization constants tests, but keep separate commits to simplify the review?

keryell commented on tests/specialization_constants/specialization_constants_same_name_inter_link.h:
```diff
@@ -0,0 +1,214 @@
+        }
+      }
+      sycl::range<1> range(1);
+      T def_value = T(value_helper<T>(0));
+      T ref_def_value = T(value_helper<T>(0));
+      T result = T(value_helper<T>(0));
+      T ref = T(value_helper<T>(0));
```
```suggestion
      T def_value { value_helper<T>(0) };
      T ref_def_value { value_helper<T>(0) };
      T result { value_helper<T>(0) };
      T ref { value_helper<T>(0) };
```

vasilytric replied:
@keryell Ronan, this was fixed in https://github.com/KhronosGroup/SYCL-CTS/pull/140

keryell commented:
> @keryell here is the link on the changes to review within this PR: [9a6354a](https://github.com/KhronosGroup/SYCL-CTS/commit/9a6354a6fd0439433f928452250c5bfeb841b0af).

Yes.
But there are some common suggestions.
Never mind, I have added some comment on this PR too.

bader commented:
> But there are some common suggestions.

That's where I think have a single pull requests with all the commits can be useful.

vasilytric commented:
@keryell @bader Alexey, Ronan

Changes from common code in another PR it's not a problem because i can refresh common code in this PR, but creating one PR with all tests (with separate commits) could be more easy for reviewers and for test developers

If need, i can move tests from https://github.com/KhronosGroup/SYCL-CTS/pull/147, https://github.com/KhronosGroup/SYCL-CTS/pull/146, https://github.com/KhronosGroup/SYCL-CTS/pull/145, https://github.com/KhronosGroup/SYCL-CTS/pull/144, https://github.com/KhronosGroup/SYCL-CTS/pull/143, https://github.com/KhronosGroup/SYCL-CTS/pull/142 to https://github.com/KhronosGroup/SYCL-CTS/pull/140

keryell commented:
It think you can keep the PR as they are and just merge https://github.com/KhronosGroup/SYCL-CTS/pull/140 in all of them and update accordingly.
It looks like some less terraforming and smaller PR as they are.

bader commented:
> It think you can keep the PR as they are and just merge #140 in all of them and update accordingly.
> It looks like some less terraforming and smaller PR as they are.

AFAIK, @kovshikovser is away. @vasilytric, can you help with addressing comments from PR #140 to speed things up?

vasilytric commented:
> > It think you can keep the PR as they are and just merge #140 in all of them and update accordingly.
> > It looks like some less terraforming and smaller PR as they are.
> 
> AFAIK, @kovshikovser is away. @vasilytric, can you help with addressing comments from PR #140 to speed things up?
@bader  Alexey, i've currently applying comments from Ronan
and wanna merge common code to my PRs after all comments will be applied

bader commented:
> > > It think you can keep the PR as they are and just merge #140 in all of them and update accordingly.
> > > It looks like some less terraforming and smaller PR as they are.
> > 
> > 
> > AFAIK, @kovshikovser is away. @vasilytric, can you help with addressing comments from PR #140 to speed things up?
> 
> @bader Alexey, i've currently applying comments from Ronan
> and wanna merge common code to my PRs after all comments will be applied

I suggest to follow Ronan's suggest i.e. address comments in PR #140, merge it and only after #140 is merge update all dependent PRs.
@vasilytric, will you be able to update https://github.com/kovshikovser/SYCL-CTS/tree/upstream_spec_const/part1 branch directly?

vasilytric commented:
> > > > It think you can keep the PR as they are and just merge #140 in all of them and update accordingly.
> > > > It looks like some less terraforming and smaller PR as they are.
> > > 
> > > 
> > > AFAIK, @kovshikovser is away. @vasilytric, can you help with addressing comments from PR #140 to speed things up?
> > 
> > 
> > @bader Alexey, i've currently applying comments from Ronan
> > and wanna merge common code to my PRs after all comments will be applied
> 
> I suggest we follow Ronan's suggestion i.e. address comments in PR #140, merge it and only after #140 is merge update all dependent PRs.
> @vasilytric, will you be able to update https://github.com/kovshikovser/SYCL-CTS/tree/upstream_spec_const/part1 branch directly?

@bader  Yes, i will update this branch: https://github.com/kovshikovser/SYCL-CTS/tree/upstream_spec_const/part1

bader commented:
@vasilytric, resolve merge conflicts with #140 in your PRs, please.

vasilytric commented:
> @vasilytric, resolve merge conflicts with #140 in your PRs, please.

@bader I will do this within 2 hours in all my PRs with merge conflicts

bader commented:
@keryell, could you take a look, please?

keryell reviewed (approved):
That looks good.
Sorry for the delay, I am back from vacation and digging into my backlog...

bader commented:
> That looks good.
> Sorry for the delay, I am back from vacation and digging into my backlog...

Welcome back. :)
Please, take a look at other pull requests adding specialization constants tests. https://github.com/KhronosGroup/SYCL-CTS/pulls
They were created ~4 weeks ago and I'm ready to hit merge button.

vasilytric commented:
@keryell Hi, Ronan, I plan to start moving try-catch blocks into test base today or tomorrow
I don't forget about this
